Add an `include` keyword

Mossaka opened the pull request:
As proposed by #169 , this PR adds an `include` keyword to the WIT.md. It also sorts the keyword list alphabetically.

alexcrichton reviewed (commented):
Could you expand a bit about the semantic meaning of the `include` statement? For example when `with` is required, what `with` does, when `include` might fail, etc.

Mossaka commented:
> Could you expand a bit about the semantic meaning of the `include` statement? For example when `with` is required, what `with` does, when `include` might fail, etc.

Good suggestion. I have thought about that but then I saw the section I've modified is for Lexical Structure. Is there a dedicated section for semantic structure of WIT? If not, I can expand the the semantic meaning in the part where I modified. @alexcrichton

alexcrichton commented:
The top of the document is probably the best place for that right now, probably around the area that `world`s are explained.

alexcrichton reviewed (commented):
Looking good, but I might recommend trying to run some examples through `wit-parser` to sort out `self`/`pkg` paths since right now they're relatively imprecise and can lead to some ambiguities in the examples I think. I realize that `include` isn't implemented yet but I think most other pieces should be able to be fleshed out to make sure they're valid.

As I think about this as well though the "de-duplication" section I think will actually need to be required for the first implementation. It's not valid to import the same interface twice under two names, so if the same interface is imported into two worlds which are union'd then it's required that a deduplication happens rather than emitting two separate imports (e.g. resolving with `with` isn't valid)

Mossaka commented:
> so if the same interface is imported into two worlds which are union'd then it's required that a deduplication happens rather than emitting two separate imports (e.g. resolving with with isn't valid)

That's a really good point. Transient deps are always hard haha. I will update the doc to include de-dep semantics.

alexcrichton commented on design/mvp/WIT.md:
```diff
@@ -185,6 +185,194 @@ default world my-world {
+    include pkg.my-world-2
+}
+
+world union-my-world-4 {
+    import a1: pkg.a
+    import b1: pkg.b
+    export c1: pkg.c
+}
```
This I think runs a risk of being surprising because `my-world-2` is effectively ignored here. Could this perhaps start out with a rule that if two interfaces are imported they must be imported under the same name? That means that the union-world above would be an error by default, but it could be fixed with `with` to do renaming to make sure the names line up.

Mossaka replied:
> if two interfaces are imported they must be imported under the same name?

that seems reasonable to me.

Mossaka replied:
I have updated the doc to address your comment.

Mossaka commented on design/mvp/WIT.md:
````diff
@@ -185,6 +185,194 @@ default world my-world {
+
+```
+
+### A Note on SubTyping
+
+As of now, host bindings are required to implicitly interpret all exports as optional to make a component targeting an included World a subtype of the union World. This [comment](https://github.com/WebAssembly/component-model/issues/169#issuecomment-1446776193) describes the reasoning behind this decision.
+
+In the future, when `optional` export is supported, the world author may explicitly mark exports as optional to make a component targeting an included World a subtype of the union World.
````
Oh I had a different interpretation of Luke's comment. I thought the goal is that we want to produce subtyped worlds and a solution to that is described by Luke as "In the short term, I expect we could get away with simply defining include to take the union of exports (just like imports) and having host bindings implicitly interpret all exports as optional"

@lukewagner please correct me if this is wrong.

alexcrichton replied:
Currently there's no support for optional exports, nor is there any planned in the near future at this time. If the goal of `include` is to produce subtyped worlds then this will need to be put on hold until optional exports are considered.

Mossaka replied:
@alexcrichton ACK, I am okay with proceeding without subtyped worlds

Mossaka replied:
I have updated the doc to address your comment.

Mossaka commented:
@alexcrichton thanks for reviewing. Could you please take another look?

alexcrichton reviewed (approved):
Looks good to me, thanks!

Mossaka commented on design/mvp/WIT.md:
```diff
@@ -185,6 +185,216 @@ default world my-world {
+// union.wit
+
+world union-my-world-1 {
+    include pkg.my-world-1
```
Yes that's right. Thanks!

Mossaka commented on design/mvp/WIT.md:
```diff
@@ -185,6 +185,216 @@ default world my-world {
+interface b { ... }
+
+// a.wit
+interface a { ... }
```
Fixed it. Thank you!

Mossaka commented:
I will rebase once #198 is merged in.

Mossaka commented:
Okay, I think this PR is ready to review @guybedford

guybedford commented on design/mvp/WIT.md:
```diff
@@ -778,6 +912,25 @@ from the root of a component and used within functions imported and exported.
 
+## Item: `include`
+
+A `include` statement enables union the current world with another world. The structure of an `include` statement is:
```
```suggestion
A `include` statement enables the union of the current world with another world. The structure of an `include` statement is:
```

guybedford commented on design/mvp/WIT.md:
````diff
@@ -778,6 +912,25 @@ from the root of a component and used within functions imported and exported.
+A `include` statement enables union the current world with another world. The structure of an `include` statement is:
+
+```wit
+include pkg.my-world-1 with { a as a1, b as b1 }
+include self.my-world-2
````
Note these are still using the dated syntax.

Mossaka replied:
Ah missed it. Thanks!

guybedford commented on design/mvp/WIT.md:
````diff
@@ -236,6 +236,139 @@ Kebab names cannot overlap and must be unique, even between imports and exports.
+}
+```
+
+### De-duplication
````
Perhaps this should go before the name conflicts section? It would also be worth clarifying how deduplication and name conflicts can be both supported. For example, can `with` syntax be used in cases where deduplication would otherwise have occured? I'd assume so but it would be good to clarify this.

Mossaka replied:
I should clarify. Deduplication only applies to IDs, and `with` syntax only applies to kebab-names.

guybedford reviewed (commented):
It would be great to see this moving forward, thanks for the update.

guybedford commented on design/mvp/WIT.md:
````diff
@@ -236,6 +236,167 @@ Kebab names cannot overlap and must be unique, even between imports and exports.
+}
+```
+
+The following example shows that `with` is a no-op when the import or export name is an ID:
````
Rather than a no-op can we just make this throw - that is, it is an error to refer to ids at all in this process? Also helps ensure any future compat concerns.

lukewagner replied:
+1

Mossaka replied:
Okay, made the `include <ID> with { ... }` case invalid.

Mossaka replied:
I am happy to send a PR to wasm-tools to adopt this decision.

lukewagner replied:
Thanks!

guybedford reviewed (approved):
LGTM

danbugs reviewed (approved):
Awesome work, Joe!

lukewagner commented on design/mvp/WIT.md:
```diff
@@ -236,6 +236,167 @@ Kebab names cannot overlap and must be unique, even between imports and exports.
+interface b { ... }
+
+// a.wit
+interface a {}
```
Also, a suggestion: with the new Wit syntax, the filenames no longer matter, so it's just as well to remove the "// a.wit" comments since they're not relevant

Mossaka replied:
Done

lukewagner commented on design/mvp/WIT.md:
```diff
@@ -778,6 +940,25 @@ from the root of a component and used within functions imported and exported.
+               | 'include' use-path 'with' '{' include-names-list '}'
+
+include-names-list ::= include-names-item
+                 | include-names-item ',' include-names-item?
```
```suggestion
                     | include-names-item ',' include-names-item?
```

lukewagner commented on design/mvp/WIT.md:
```diff
@@ -236,6 +236,158 @@ Kebab names cannot overlap and must be unique, even between imports and exports.
+
+### Name Conflicts and `with`
+
+When two or more included Worlds have the same name for an import or export, the name is considered to be in conflict. The conflict needs to be explicitly resolved by the world author using the `with` keyword. `with` allows the world author to rename the import or export to a different name.
+
+Notice that when import or export names are IDs and since IDs are unique, there is no need to resolve name conflicts. Thus the `with` syntax is invalid when used with `include <ID>`. Only when import or export names are kebab names, name conflicts need to be resolved.
+
```
```suggestion
When two or more included Worlds have the same name for an import or export that does *not* have an ID, de-duplication cannot be used (because the two same-named imports/exports might have different meanings in the different worlds) and thus the conflict has to be resolved using the `with` keyword:
```

lukewagner commented on design/mvp/WIT.md:
````diff
@@ -236,6 +236,158 @@ Kebab names cannot overlap and must be unique, even between imports and exports.
+}
+```
+
+The following example shows an invalid example that `with` is used when the import or export name is an ID:
+
+```wi
+package local:demo
+
+world my-world-a {
+    import a1
+    import b1
+}
+
+world my-world-b {
+    import a1
+    import b1
+}
+
+world union-my-world-a {
+    include my-world-a with { a1 as a3 }
+    include my-world-b with { a1 as a2 }
+}
+```
````
```suggestion
`with` cannot be used to rename IDs, however, so the following world would be invalid:
```wit
package local:demo

interface a {
  foo: func()
}

world world-using-a {
    import a
}

world invalid-union-world {
    include my-using-a with { a as b }  // invalid: 'a', which is short for 'local:demo/a', is an ID
}
```
```

lukewagner reviewed (commented):
Thanks for the updates, and great work.  Regarding the most recent change, which says that you can't rename an ID, unfortunately we can't just rule it out syntactically, because `include x with {a as b }` may or may not be allowed depending on `x`.  Here's a suggested rewording, but feel free to reject and do it your own way.

Mossaka commented:
Thank you, @lukewagner !! I have committed your suggestions.

lukewagner reviewed (approved):
Thanks Joe, overall lgtm, thanks for spearheading this!  I'm thinking I'll merge early next week if there aren't any more comments from anyone.

Mossaka commented:
Thanks everyone for reviewing this!! This meant a lot to me as it marks the first contribution I've contributed to a Wasm proposal. Cheers 🙌 🥂
